fix: clear the discovery cache after CRDs are installed

technosophos opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/master/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:
This fixes a bug in which CRD instances could not be created in the same chart as the CRDs themselves. This invalidates the local discovery cache and forces a round-trip discovery operation with the Kubernetes API server.

**Special notes for your reviewer**:

Closes #6316 

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

technosophos committed:
fix: clear the discovery cache after CRDs are installed

This fixes an issue in which a chart could not contain both a CRD and an instance of that CRD. It works around a stale cache by force cache invalidation whenever a CRD is added.

Closes #6316

Signed-off-by: Matt Butcher <matt.butcher@microsoft.com>

hickeyma commented:
@thequailman can you test if this fixes the issue?

karuppiah7890 commented:
I'm wondering if it's possible to write a failing test for the issue and then pass it with this PR

thomastaylor312 reviewed (changes requested):
I can't get this to work for me and I think I know why, but am not sure how to fix it. We should pair and see what we can come up with

technosophos committed:
fix: wait for CRD to register before allowing CRDs to be installed

This fixes an issue with the previous version of this patch in which the CRD would not be available quickly enough.

Signed-off-by: Matt Butcher <matt.butcher@microsoft.com>

technosophos commented:
Looks like were two issues with the earlier draft of this patch:

- The capabilities cache gets generated before the CRDs are installed. I reversed the order of this.
- Some clusters take longer to register CRDs. I was testing locally with Kind, which is fast. But some take a second or two.

The following script is what I have been using for testing:

```bash
BASPATH=/Users/technosophos/Code/Go/src/helm.sh/helm
HELM=$BASPATH/bin/helm
CHART=$BASPATH/_scratch/test-0.1.0.tgz

# Build a fresh Helm
echo "Building Helm"
cd $BASPATH && make build

echo "Cleaning environment"
# Clear all CRDs
kubectl delete crds --all
# Clear cache
rm -r /Users/technosophos/.kube/cache/discovery/127.0.0.1_*

echo "Running test"
# Run a command that forces discovery client to regen
# Install Helm chart
$HELM install --debug test1 $CHART

echo "Deleting helm install"
$HELM delete test1
```

thomastaylor312 commented on pkg/action/install.go:
```diff
@@ -111,6 +111,49 @@ func (i *Install) Run(chrt *chart.Chart, vals map[string]interface{}) (*release.
+				i.cfg.Log("Clearing discovery cache")
+				discoveryClient.Invalidate()
+				// Give time for the CRD to be recognized.
+				time.Sleep(5 * time.Second)
```
We should probably clean up and reuse the code found here in Helm 2 so we aren't just sleeping and hoping it works: https://github.com/helm/helm/blob/3ed073b2e0c37dd5c994562d6fe8e04b5f3609eb/pkg/kube/client.go#L584-L612

technosophos replied:
I am adding support to Wait() for CRDs, and then just using that. It has some hidden advantages for if people try to sneak non-CRD things into the CRDs() directory. But _sssssh... don't tell them that they can do this_.

thomastaylor312 replied:
![tenor-138232708](https://user-images.githubusercontent.com/4294228/64203015-d1da6580-ce4f-11e9-82f2-4c6487485beb.gif)

technosophos replied:
@thomastaylor312 Okay, this is now done. We're using an updated version of the Wait logic to watch for the CRD to become available.

thomastaylor312 reviewed (commented):
I think we already have some code we can copy paste instead of sleeping to wait for CRDs to reach an established state

technosophos committed:
feat: use Wait() to wait for CRDs to be ready

This forward-ports the CRD wait logic to Helm 3, and then uses that to wait for CRDs to be registered.

Signed-off-by: Matt Butcher <matt.butcher@microsoft.com>

technosophos commented on pkg/kube/wait.go:
```diff
@@ -46,6 +48,9 @@ type waiter struct {
 // until all are ready or a timeout is reached
 func (w *waiter) waitForResources(created ResourceList) error {
 	w.log("beginning wait for %d resources with timeout of %v", len(created), w.timeout)
+	if err := apiextv1beta1.AddToScheme(scheme.Scheme); err != nil {
```
Adding CRDs to the globally shared Scheme (Yay, Singletons!) is probably best done elsewhere. Any idea where?

technosophos committed:
ref: moved the scheme modification to an appropriate place.

Signed-off-by: Matt Butcher <matt.butcher@microsoft.com>

thomastaylor312 commented on pkg/action/install.go:
```diff
@@ -103,6 +103,50 @@ func NewInstall(cfg *Configuration) *Install {
+		if discoveryClient, err := i.cfg.RESTClientGetter.ToDiscoveryClient(); err != nil {
+			// On error, we don't want to bail out, since this is unlikely
+			// to impact the majority of charts.
+			i.cfg.Log("Could not clear the discovery cache: %s", err)
```
So I was thinking about this some more. I think we actually _do_ want to bail out here. If we got into this block, the user was meaning to install CRDs and we don't want their release to just fail on them. What do you think?

thomastaylor312 replied:
Oh and also bail out on a failed wait

technosophos replied:
I don't like the idea of failing on transient errors that are unlikely to impact the user. It seems wrong to me to take something that is not fatal (a cleared cache, a wait that is likely not going to take long anyway) and then require the user to take action.

However, I don't care enough to drag on this PR longer, so I will change it.

thomastaylor312 commented on pkg/kube/client.go:
```diff
@@ -55,6 +58,10 @@ func New(getter genericclioptions.RESTClientGetter) *Client {
 		getter = genericclioptions.NewConfigFlags(true)
 	}
+	if err := apiextv1beta1.AddToScheme(scheme.Scheme); err != nil {
+		// This should never happen.
```
Famous last words... 😉

thomastaylor312 commented on pkg/kube/client.go:
```diff
@@ -55,6 +58,10 @@ func New(getter genericclioptions.RESTClientGetter) *Client {
 	if getter == nil {
 		getter = genericclioptions.NewConfigFlags(true)
 	}
+	if err := apiextv1beta1.AddToScheme(scheme.Scheme); err != nil {
```
We should add a comment here so people don't have to do the same research you did

thomastaylor312 commented on pkg/kube/wait.go:
```diff
@@ -257,6 +270,26 @@ func (w *waiter) daemonSetReady(ds *appsv1.DaemonSet) bool {
+		case apiextv1beta1.NamesAccepted:
+			if cond.Status == apiextv1beta1.ConditionFalse {
+				// This indicates a naming conflict, but it's probably not the
+				// job of this function to faile because of that. Instead,
```
Spelling nit: s/faile/fail

thomastaylor312 reviewed (changes requested):
Just a few small changes that I found on this second pass

thomastaylor312 commented on pkg/action/install.go:
```diff
@@ -103,6 +103,50 @@ func NewInstall(cfg *Configuration) *Install {
+			i.cfg.Log("Could not clear the discovery cache: %s", err)
+		} else {
+			i.cfg.Log("Clearing discovery cache")
+			discoveryClient.Invalidate()
```
I was looking through this one more time and I think we should do the cache invalidation at the very end, outside of the for loop. When I looked into the code, I think the discovery client calls the api multiple times for each api group, meaning that if someone installs 5-10 CRDs, they will be calling their API hundreds of times

thomastaylor312 replied:
Source: https://github.com/kubernetes/cli-runtime/blob/9d670f2021f41f32ed762a6ddbc1c3ac7c441126/pkg/genericclioptions/config_flags.go#L222-L225

technosophos replied:
So originally I was doing this so that the crds folder could have things that depended on each other. And this operates _per file_, not per CRD. But I'll switch it.

technosophos committed:
fix: turned warnings into fatal errors, fixed spelling, clear cache once

Signed-off-by: Matt Butcher <matt.butcher@microsoft.com>

technosophos commented:
Please re-review.

thomastaylor312 reviewed (approved):
LGTM
